### PR TITLE
Update state.py

### PR DESCRIPTION
**Pull request by jubrad**
### What does this PR do?
Updates `salt/state.py`
create `_clean_tag()` method to safely encode tag names for file creation
use `_clean_tag()` method  for both create and read of state cache file

### What issues does this PR fix or reference?
#42706

### Previous Behavior
used _gen_tag() method for cache file name which could use invalid characters for file open calls.

### New Behavior
urllib.quote is used with the `safe=''` option to escape characters which may cause file creation to fail.

### Tests written?

No( will do if asked )

This has been tested manually on a centos7 docker container, but I do not have the resources to test on windows. I can also look into writing or modifying a test for this if necessary.

### CODE REVIEW

**Comment by damon-atkins** [comment]
I would suggest an alternative would be to limit the tag names to something sensible and supported on all file systems e.g.
```
 if re.match(u'[.a-zA-Z0-9_]+$',tag,flags=re.UNICODE) is None:
      raise error bad tag name
```

Use \w if you wish to allow non-ASCII chars.

This way someone will not code into salt a bad tag name to being with.

**Comment by jubrad** [comment]
@damon-atkins 

```
# test_parallel.sls
/test:
  file.managed:
    - contents:
      - it's working
    - parallel: True

m2:
  cmd.run:
    - name: echo '/'
    - parallel: True
```

calling `salt-call state.apply test_parallel` will generate the following cache files.
Note: it does not leave the `/` char. 

```bash
[root@5f3ce76caa67 /]# ls /var/cache/salt/minion/20170817161223630299
cmd_%7C-m2_%7C-echo%20%27%2F%27_%7C-run  file_%7C-%2Ftest_%7C-%2Ftest_%7C-managed
```

That being said I did not know about `salt.utils.sanitize_win_path_string` and will look into it to see if it is a generically better solution.

**Comment by jubrad** [comment]
I could always have done something that simply removed characters which caused issues, but this seemed like it had the potential to cause cache file collision. I believe salt needs to have unique tags so finding something that could be encoded from a tag and decoded back to the tag seemed like the optimal solution; as such, I do not think `salt.utils.sanitize_win_path_string` will be a good solution for this problem..

**Comment by terminalmage** [comment]
Since the referenced bug was reported against 2017.7.0, and that's when the parallel execution feature was added, this will need a backport.

**Comment by damon-atkins** [comment]
Salt is heading to all strings being Unicode in PY2. Looks like quote

```
>>> a=u'0unicode_file_name_üöÜä.txt'
>>> print quote('0unicode_file_name_üöÜä.txt')
>>> print unquote('0unicode_file_name_%C3%BC%C3%B6%C3%9C%C3%A4.txt')
0unicode_file_name_üöÜä.txt
>>> a
u'0unicode_file_name_\xfc\xf6\xdc\xe4.txt'
>>> print quote(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.6/urllib.py", line 1223, in quote
    res = map(safe_map.__getitem__, s)
KeyError: u'\xfc'
>>> print quote(a,safe='')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.6/urllib.py", line 1223, in quote
    res = map(safe_map.__getitem__, s)
KeyError: u'\xfc'
>>>
```
This needs to work with Unicode Strings in PY2.

I think the salt team prefer 
`import urllib` over from `urllib import quote`

I suggest adding to salt.utils or salt.utils.files something like this and then removing salt.utils.sanitize_win_path_string
```
def safe_filename_leaf(basename):
    '''
    input the basename of a file, without the directory tree, and returns a safe name to use
    '''
    return re.sub(u'[\\\/:*?"<>|]',u'_',basename,flags=re.UNICODE)

def safe_filepath(file_path_name):
    '''
    input the full path and filename, splits on directory separator and calls safe_filename_leaf for each part of the path, Note c:\ would become c_\
    '''
   (drive,path)=os.path.splitdrive(file_path_name)
   path=os.sep.join([safe_filename_leaf(x) for x in file_path_name.rsplit(os.sep)])
   if drive:
       return os.sep.join([drive,path])
   else:
       return path
```

Maybe check os.path.supports_unicode_filenames

**Comment by jubrad** [comment]
@damon-atkins
Great catch on the unicode problem. 

The proposed regex`re.sub(u'[\\\/:*?"<>|]',u'_',basename,flags=re.UNICODE)`  will not work. 

Example of issue:
```salt
test '_' = '*':
  cmd.run:
    - parallel: True

test '_' = '*':
  cmd.run:
    - parallel: True
```
With urllib.quote:
```bash
[root@5f3ce76caa67 /]# salt-call state.apply test_conflict
[root@5f3ce76caa67 /]# ls /var/cache/salt/minion/20170818142804165329
cmd_%7C-test%20%27%2A%27%20%3D%20%27_%27_%7C-test%20%27%2A%27%20%3D%20%27_%27_%7C-run
cmd_%7C-test%20%27_%27%20%3D%20%27%2A%27_%7C-test%20%27_%27%20%3D%20%27%2A%27_%7C-run
```
If I used the proposed regex(note only one file is created):
```bash
[root@5f3ce76caa67 /]# ls /var/cache/salt/minion/20170818154956709238
cmd__-test '_' = '_'__-test '_' = '_'__-run
```

I still think urllib.quote is the correct answer, but an encoding to utf8 encoding first.
```
>>> quote(u'<>:"\|?*üöÜä'.encode('utf8'),safe='')
'%3C%3E%3A%22%5C%7C%3F%2A%C3%BC%C3%B6%C3%9C%C3%A4'
```

This also seems to meet microsoft and linux file name reqs except for length.
https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx

Also the caching scheme and tag generation seem to be relatively contained to this file so I don't see why this solution should be pushed into utils.

**Comment by garethgreenaway** [comment]
re-run lint

File: salt/state.py [hunk 1]
```diff
@@ -149,6 +150,13 @@ def _gen_tag(low):
     return u'{0[state]}_|-{0[__id__]}_|-{0[name]}_|-{0[fun]}'.format(low)
 
 
+def _clean_tag(tag):
+    '''
+      urllib safe quote the tag value to avoid invalid chars in the filename
+    '''
+    return urllib.quote(tag.encode('utf8'), safe='')
```
Comment (cachedout) [inline salt/state.py]: I don't think we want to force this to `UTF-8`. Could you instead use `__salt_system_encoding__`?

Comment (jubrad) [inline salt/state.py]: Yeah I'm actually planning on using this method:
https://github.com/saltstack/salt/pull/43056/files#diff-bdabb2b499cd3b4adf29d764d4eec419R168

Comment (cachedout) [reply salt/state.py]: Sounds good. If you can change this then we can re-review this. Thanks.

**Comment by damon-atkins** [comment]
Can this be closed?

**Comment by cachedout** [comment]
@damon-atkins Why should it be closed? The last comment was that additional changes were coming.

**Comment by damon-atkins** [comment]
Because new 2 PR have replaced it.

**Comment by cachedout** [comment]
@damon-atkins Can you please link me to the PRs you're referring to?

**Comment by damon-atkins** [comment]
See https://github.com/saltstack/salt/pull/43018#ref-pullrequest-252374553 `Bp 43018 #43159`

**Comment by cachedout** [comment]
Ah, OK. I understand now. Yes, let's close this.

**Comment by jubrad** [comment]
@cachedout @damon-atkins Sorry I've been away and haven't had time to update in accordance to my last comment. 

As I understand it the back port is in; however, this issue will still exist in develop. If this is to be fixed in develop I'll need the code introduced into `2017.7.2` in #43056 to be merged into develop and then I can adjust this pr to use that code. 

If back ported code from released branches is cycled back into develop periodically then this pr should most certainly be closed, but I see no signs of that happening and it's not stated in the contribution document.

**Comment by cachedout** [comment]
Hi @jubrad That code should have already been merged forward from that bug-fix branch into develop. 

@rallytime could you review the contribution document and see if there are any changes we need to make to clarify this? Thanks.

**Comment by rallytime** [comment]
Yes, the PR mentioned above has already been merged into develop. The "merge-forward" process is documented [here](https://docs.saltstack.com/en/latest/topics/development/contributing.html#the-current-release-branch), but I don't think those have been revised for quite some time.

Now would be a good time to review these and make clarifications as we are moving towards using even more release branches.

**Pull request closed without merging**